platform/ScopedLock.h

```diff
@@ -0,0 +1,87 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SCOPEDLOCK_H
+#define MBED_SCOPEDLOCK_H
+
+#include "platform/NonCopyable.h"
+
+namespace mbed {
+
+/** \addtogroup platform */
+/** @{*/
+/**
+ * \defgroup platform_ScopedLock ScopedLock functions
+ * @{
+ */
+
+/** RAII-style mechanism for owning a lock of Lockable object for the duration of a scoped block
+ *
+ * @tparam Lockable The type implementing BasicLockable concept
+ *
+ * @note For type Lockable to be BasicLockable the following conditions have to be satisfied:
+ *        - has public method @a lock which blocks until a lock can be obtained for the current execution context
```

Comment:
Nitpick: There is no such thing as a *method* in C++; these are named member function.

Reply:
If we're being that nitpicky, it doesn't actually have to be a member function. `lock` could be a function pointer or a `mbed::Callback`, or maybe something else I can't think of.

The C++ standard just talks about the expression `m.lock()` (if `m` is a value of a `BasicLockable` type).

(I guess we can't just quote the C++ standard in documentation? Or cop out by referring to it?)

Reply:
Guys our reviews tend to stretch to infinity... Does using `method` instead of `named member function` or quoting part of the C++ standard introduces unwanted ambiguity? I mean if it's one of many things to change by all means, but if it's the only thing maybe we can agree it's close enough and try to get it merged and focus on next task?

Reply:
@pan- started it :) 

Nitpicking is silly - that's why I was wondering if we should in general just reference the C++ standard for this sort of thing so we don't have to spend time coming up with our own probably-slightly-off paraphrasing. It's already in the standard or at http://en.cppreference.com/w/cpp/concept/BasicLockable. Not sure whether our documentation style requires us to be entirely self-contained.